tracer/src/Datadog.Trace/Ci/Agent/CIAgentlessWriter.cs

```diff
@@ -26,28 +26,30 @@ internal sealed class CIAgentlessWriter : IEventWriter
         private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<CIAgentlessWriter>();
 
         private readonly BlockingCollection<IEvent> _eventQueue;
-        private readonly TaskCompletionSource<bool> _flushTaskCompletionSource;
-        private readonly Task _periodicFlush;
         private readonly AutoResetEvent _flushDelayEvent;
+        private readonly Buffers[] _buffersArray;
 
-        private readonly EventsPayload _ciTestCycleBuffer;
-
-        private readonly ICIAgentlessWriterSender _sender;
-
-        public CIAgentlessWriter(ICIAgentlessWriterSender sender)
+        public CIAgentlessWriter(ICIAgentlessWriterSender sender, IFormatterResolver formatterResolver = null, int? concurrency = null)
         {
             _eventQueue = new BlockingCollection<IEvent>(MaxItemsInQueue);
-            _flushTaskCompletionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
             _flushDelayEvent = new AutoResetEvent(false);
 
-            _ciTestCycleBuffer = new CITestCyclePayload();
-
-            _sender = sender;
-
-            _periodicFlush = Task.Factory.StartNew(InternalFlushEventsAsync, this, TaskCreationOptions.LongRunning);
-            _periodicFlush.ContinueWith(t => Log.Error(t.Exception, "Error in sending ciapp events"), TaskContinuationOptions.OnlyOnFaulted);
+            // Concurrency Level is a simple algorithm where we select a number between 1 and 8 depending on the number of Logical Processor Count
+            // To scale the number of senders with a hard limit.
+            var concurrencyLevel = concurrency ?? Math.Min(Math.Max(Environment.ProcessorCount / 2, 1), 8);
+            _buffersArray = new Buffers[concurrencyLevel];
```

Comment:
Why do we need more than one payload now and we didn't before? Where do we chose which of the N `Buffers` object will get used for a given event/span/whatever?

Reply:
Ok, to answer that I have to explain the current architecture of the writer. Nice ASCII art incoming...

```
        ┌────────────────────────────────────────────────────────────────┐
        │                                                                │
        │ CIAgentlessWriter         ┌────────────────────────────────┐   │
        │                           │ Buffers                        │   │
        │                         ┌─┤                                │   │
        │                         │ │ ┌────────────────────────────┐ │   │
        │                         │ │ │ CITestCyclePayload  Buffer │ │   │
        │                         │ │ │                            │ │   │
        │                         │ │ │         SpanEvent          │ │   │  CITestCyclePayload Url
        │                         │ │ │         TestEvent          ├─┼───┼────────────────────────────►
        │   ┌───────────────┐     │ │ ├────────────────────────────┤ │   │
        │   │               │     │ │ │ Items: inf     Bytes: 4MB  │ │   │
        │   │  Event Queue  │     │ │ └────────────────────────────┘ │   │
IEvent  │   │               │     │ │                                │   │
 ───────┼──►│               ├─────┤►│ ┌────────────────────────────┐ │   │
        │   │   Max: 2500   │     │ │ │ CICodeCoveragePayload Buf. │ │   │
        │   │               │     │ │ │                            │ │   │
        │   └───────────────┘     │ │ │      CoveragePayload       │ │   │  CICodeCoveragePayload Url
        │                         │ │ │                            ├─┼───┼────────────────────────────►
        │                         │ │ ├────────────────────────────┤ │   │
        │                         │ │ │ Items: 100     Bytes: 50MB │ │   │
        │                         │ │ └────────────────────────────┘ │   │
        │                         │ │                                │   │
        │                         │ │ Flush each sec or limit reach  │   │
        │                         │ └──────────────────────────────┬─┘   │
        │                         │                                │     │
        │                         └────────────────────────────────┘     │
        │                                   1 .. N Consumers             │
        │                                                                │
        │                                      Max N = 8                 │
        │                                                                │
        └────────────────────────────────────────────────────────────────┘
```

The agentless writer receives `IEvent` instances, those events are stored in a bounded `BlockingCollection` with a max of 2500 events where we use a Producer/Consumer mechanism over the collection. That collection is what we call the EventQueue. If the collection reach the max count, we stop the ingestion of new events (blocking the thread) until we process those events and the number decreases (we cannot drop any event).

For each backend endpoint we have a Payload buffer (`CITestCyclePayload` and `CICodeCoveragePayload`) , is different for each endpoint because the endpoint not only have different urls, but also different encoding and different limits for each payload.

A `Buffer` instance contains the two payload buffers and a thread that consumes an item from the queue, ask to each payload buffer which one can process the type of event and then add the item to that buffer. The item will be serialized to the format expected by the endpoint. A flush happens when the timeout period occurs (Currently 1 second) and the payload buffer contains any event, or when the payload buffer is full (this will depend on the limits of each payload).

**Current limits are:**
- `CITestCyclePayload`: Number of items is not bounded but payload size must be < 5MB
-  `CICodeCoveragePayload`: Max items per payload = 100, and Max Bytes per Payload = 50MB

Before this PR, we had 1 single instance of the buffer with 1 single thread consuming all events from the queue. And that could work in this case as well, but because the limitation of the code coverage payload I could observe a bottleneck in the writer when using a single thread.

For example, when running over the DuckTyping test suite with > 3000 tests the logs showed something similar to...
```
Sending 1900 events....
Sending 100 coverages
Sending 100 coverages
Sending 100 coverages
Sending 100 coverages
....
```
Having here a single thread means that for sending other coverage payload the buffer needs to wait that the previous request was successful.

That's why I introduce the concurrency concept in this PR, now we can have multiple consumers over the same queue and we can parallelize sending data to the same endpoint.

Reply:
Makes sense. The only limitation I find to the current architecture is that we must have the same amount of `CICodeCoveragePayload` and `CITestCyclePayload` and while being able to send more code coverage payloads at a time does improve performance we didn't really need this for test cycle payloads.

Reply:
Maybe it's short-sighted, but I feel like `Buffers` could have two implementations, one for `CITestCyclePayload` and one for `CICodeCoveragePayload`. We could then have a single `CITestCyclePayloadBuffer` and max 7 CICodeCoveragePayload? 🤔  Might simplify some of the flushing logic in `InternalFlushEventsAsync` then, as there's ne need for the `Task.WhenAll`s and all the "duplicate" calls?